File: modules/sozia-mediapipe/ios/SoziaMediaPipeHolistic.swift
import Foundation
import MediaPipeTasksVision

/// Holds the three MediaPipe landmarkers and runs serial inference per frame.
/// All public methods are safe to call from any thread; `load()` must be
/// dispatched onto a background queue by the caller.
final class SoziaMediaPipeHolistic {

  // MARK: - State

  private var faceLandmarker: FaceLandmarker?
  private var handLandmarker: HandLandmarker?
  private var poseLandmarker: PoseLandmarker?
  private(set) var isReady = false

  // MARK: - Face index filter (matches FACE_LANDMARK_INDICES in TypeScript)

  private static let faceIndices: [Int] = [
    0, 1, 4, 5, 13, 14, 17, 33, 37, 39, 40, 46, 52, 53, 55, 61, 65,
    78, 80, 81, 82, 84, 87, 88, 91, 95, 133, 144, 145, 146, 152, 157,
    158, 159, 160, 175, 178, 181, 185, 191, 199, 200, 263, 267, 269,
    270, 276, 282, 283, 285, 291, 295, 308, 310, 311, 312, 314, 317,
    318, 321, 324, 362, 373, 374, 375, 384, 385, 386, 387, 402, 405,
    409, 415, 468, 469, 470, 471, 472, 473, 474, 475, 476, 477,
  ]

  // MARK: - Result type

  struct LandmarkResult {
    let faceLandmarks: [[Float]]?      // 83×3
    let leftHandLandmarks: [[Float]]?  // 21×3
    let rightHandLandmarks: [[Float]]? // 21×3
    let poseLandmarks: [[Float]]?      // 33×3
  }

  // MARK: - Model loading

  /// Loads the three .task bundles from the main bundle.
  /// Expected asset names: `face_landmarker.task`, `hand_landmarker.task`, `pose_landmarker_full.task`.
  /// Sets `isReady = true` on success; leaves it false on any error.
  func load() {
    guard
      let facePath  = Bundle.main.path(forResource: "face_landmarker",      ofType: "task"),
      let handPath  = Bundle.main.path(forResource: "hand_landmarker",      ofType: "task"),
      let posePath  = Bundle.main.path(forResource: "pose_landmarker_full", ofType: "task")
    else {
      return
    }

    do {
      let faceOptions = FaceLandmarkerOptions()
      faceOptions.baseOptions.modelAssetPath = facePath
      faceOptions.runningMode = .video
      faceOptions.numFaces = 1
      faceOptions.minFaceDetectionConfidence = 0.3
      faceOptions.minFacePresenceConfidence  = 0.3
      faceLandmarker = try FaceLandmarker(options: faceOptions)

      let handOptions = HandLandmarkerOptions()
      handOptions.baseOptions.modelAssetPath = handPath
      handOptions.runningMode = .video
      handOptions.numHands = 2
      handOptions.minHandDetectionConfidence = 0.3
      handOptions.minHandPresenceConfidence  = 0.3
      handLandmarker = try HandLandmarker(options: handOptions)

      let poseOptions = PoseLandmarkerOptions()
      poseOptions.baseOptions.modelAssetPath = posePath
      poseOptions.runningMode = .video
      poseOptions.numPoses = 1
      poseLandmarker = try PoseLandmarker(options: poseOptions)

      isReady = true
    } catch {
      isReady = false
    }
  }

  // MARK: - Inference

  func detect(sampleBuffer: CMSampleBuffer, timestampMs: Int) -> LandmarkResult? {
    guard isReady,
          let face = faceLandmarker,
          let hand = handLandmarker,
          let pose = poseLandmarker
    else { return nil }

    guard let image = try? MPImage(sampleBuffer: sampleBuffer) else { return nil }

    let ts = TimeInterval(timestampMs)

    guard let faceResult = try? face.detectAsync(image: image, timestampInMilliseconds: Int(ts)),
          let handResult = try? hand.detectAsync(image: image, timestampInMilliseconds: Int(ts)),
          let poseResult = try? pose.detectAsync(image: image, timestampInMilliseconds: Int(ts))
    else { return nil }

    let faceLandmarks = filterFaceLandmarks(faceResult.faceLandmarks.first)
    let (leftHand, rightHand) = partitionHands(handResult)
    let poseLandmarks = mapLandmarks(poseResult.landmarks.first)

    return LandmarkResult(
      faceLandmarks: faceLandmarks,
      leftHandLandmarks: leftHand,
      rightHandLandmarks: rightHand,
      poseLandmarks: poseLandmarks
    )
  }

  // MARK: - Private helpers

  private func filterFaceLandmarks(_ landmarks: [NormalizedLandmark]?) -> [[Float]]? {
    guard let all = landmarks, !all.isEmpty else { return nil }
    return Self.faceIndices.compactMap { idx in
      guard idx < all.count else { return nil }
      let lm = all[idx]
      return [lm.x, lm.y, lm.z]
    }
  }

  private func mapLandmarks(_ landmarks: [NormalizedLandmark]?) -> [[Float]]? {
    guard let all = landmarks, !all.isEmpty else { return nil }
    return all.map { lm in [lm.x, lm.y, lm.z] }
  }

  private func partitionHands(_ result: HandLandmarkerResult) -> ([[Float]]?, [[Float]]?) {
    var left: [[Float]]?
    var right: [[Float]]?
    for (i, handedness) in result.handedness.enumerated() {
      guard let label = handedness.first?.categoryName?.lowercased(),
            i < result.landmarks.count else { continue }
      let mapped = mapLandmarks(result.landmarks[i])
      if label == "left",  left  == nil { left  = mapped }
      if label == "right", right == nil { right = mapped }
    }
    return (left, right)
  }
}

File: modules/sozia-mediapipe/android/src/main/java/com/sozia/mediapipe/SoziaMediaPipeHolistic.kt
package com.sozia.mediapipe

import android.content.Context
import com.google.mediapipe.tasks.core.BaseOptions
import com.google.mediapipe.tasks.vision.core.RunningMode
import com.google.mediapipe.tasks.vision.facelandmarker.FaceLandmarker
import com.google.mediapipe.tasks.vision.handlandmarker.HandLandmarker
import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarker
import com.google.mediapipe.framework.image.MPImage
import com.google.mediapipe.tasks.components.containers.NormalizedLandmark

/**
 * Runs FaceLandmarker, HandLandmarker, and PoseLandmarker serially.
 * Models are read from assets/mediapipe/ on first call to load().
 */
class SoziaMediaPipeHolistic(private val context: Context) {

    private var faceLandmarker: FaceLandmarker? = null
    private var handLandmarker: HandLandmarker? = null
    private var poseLandmarker: PoseLandmarker? = null

    var isReady = false
        private set

    companion object {
        // Matches FACE_LANDMARK_INDICES in TypeScript (sozia/common/models/index.ts)
        val FACE_INDICES = intArrayOf(
            0, 1, 4, 5, 13, 14, 17, 33, 37, 39, 40, 46, 52, 53, 55, 61, 65,
            78, 80, 81, 82, 84, 87, 88, 91, 95, 133, 144, 145, 146, 152, 157,
            158, 159, 160, 175, 178, 181, 185, 191, 199, 200, 263, 267, 269,
            270, 276, 282, 283, 285, 291, 295, 308, 310, 311, 312, 314, 317,
            318, 321, 324, 362, 373, 374, 375, 384, 385, 386, 387, 402, 405,
            409, 415, 468, 469, 470, 471, 472, 473, 474, 475, 476, 477
        )
    }

    data class LandmarkResult(
        val faceLandmarks: List<List<Float>>?,
        val leftHandLandmarks: List<List<Float>>?,
        val rightHandLandmarks: List<List<Float>>?,
        val poseLandmarks: List<List<Float>>?
    )

    fun load() {
        try {
            val faceOptions = FaceLandmarker.FaceLandmarkerOptions.builder()
                .setBaseOptions(BaseOptions.builder().setModelAssetPath("mediapipe/face_landmarker.task").build())
                .setRunningMode(RunningMode.VIDEO)
                .setNumFaces(1)
                .setMinFaceDetectionConfidence(0.3f)
                .setMinFacePresenceConfidence(0.3f)
                .build()
            faceLandmarker = FaceLandmarker.createFromOptions(context, faceOptions)

            val handOptions = HandLandmarker.HandLandmarkerOptions.builder()
                .setBaseOptions(BaseOptions.builder().setModelAssetPath("mediapipe/hand_landmarker.task").build())
                .setRunningMode(RunningMode.VIDEO)
                .setNumHands(2)
                .setMinHandDetectionConfidence(0.3f)
                .setMinHandPresenceConfidence(0.3f)
                .build()
            handLandmarker = HandLandmarker.createFromOptions(context, handOptions)

            val poseOptions = PoseLandmarker.PoseLandmarkerOptions.builder()
                .setBaseOptions(BaseOptions.builder().setModelAssetPath("mediapipe/pose_landmarker_full.task").build())
                .setRunningMode(RunningMode.VIDEO)
                .setNumPoses(1)
                .build()
            poseLandmarker = PoseLandmarker.createFromOptions(context, poseOptions)

            isReady = true
        } catch (e: Exception) {
            isReady = false
        }
    }

    fun detect(image: MPImage, timestampMs: Long): LandmarkResult? {
        if (!isReady) return null

        val faceResult = faceLandmarker?.detectForVideo(image, timestampMs) ?: return null
        val handResult = handLandmarker?.detectForVideo(image, timestampMs) ?: return null
        val poseResult = poseLandmarker?.detectForVideo(image, timestampMs) ?: return null

        val allFace: List<NormalizedLandmark>? = faceResult.faceLandmarks().firstOrNull()
        // Always emit exactly 83 entries; zero-fill any index beyond the mesh size.
        // Matches the web backend convention and prevents server validation errors on partial occlusion.
        val faceLandmarks: List<List<Float>>? = if (allFace != null) {
            FACE_INDICES.toList().map { idx: Int ->
                if (idx < allFace.size) {
                    val lm: NormalizedLandmark = allFace[idx]
                    listOf(lm.x(), lm.y(), lm.z())
                } else listOf(0f, 0f, 0f)
            }
        } else null

        var leftHand: List<List<Float>>? = null
        var rightHand: List<List<Float>>? = null
        for (i in handResult.handednesses().indices) {
            val label = handResult.handednesses()[i].firstOrNull()?.categoryName()?.lowercase()
            val handPoints: List<NormalizedLandmark> = handResult.landmarks()[i]
            val mapped: List<List<Float>> = handPoints.map { lm: NormalizedLandmark ->
                listOf(lm.x(), lm.y(), lm.z())
            }
            when {
                label == "left"  && leftHand  == null -> leftHand  = mapped
                label == "right" && rightHand == null -> rightHand = mapped
            }
        }

        val posePoints: List<NormalizedLandmark>? = poseResult.landmarks().firstOrNull()
        val poseLandmarks: List<List<Float>>? = posePoints?.map { lm: NormalizedLandmark ->
            listOf(lm.x(), lm.y(), lm.z())
        }

        return LandmarkResult(faceLandmarks, leftHand, rightHand, poseLandmarks)
    }
}
